--- a/src/mcp_proxy/proxy_server.py
+++ b/src/mcp_proxy/proxy_server.py
@@ -25,8 +25,8 @@
     if capabilities.prompts:
         logger.debug("Capabilities: adding Prompts...")
 
-        async def _list_prompts(_: t.Any) -> types.ServerResult:  # noqa: ANN401
-            result = await remote_app.list_prompts()
+        async def _list_prompts(req: types.ListPromptsRequest) -> types.ServerResult:
+            result = await remote_app.list_prompts(params=req.params)
             return types.ServerResult(result)
 
         app.request_handlers[types.ListPromptsRequest] = _list_prompts
@@ -40,14 +40,16 @@
     if capabilities.resources:
         logger.debug("Capabilities: adding Resources...")
 
-        async def _list_resources(_: t.Any) -> types.ServerResult:  # noqa: ANN401
-            result = await remote_app.list_resources()
+        async def _list_resources(req: types.ListResourcesRequest) -> types.ServerResult:
+            result = await remote_app.list_resources(params=req.params)
             return types.ServerResult(result)
 
         app.request_handlers[types.ListResourcesRequest] = _list_resources
 
-        async def _list_resource_templates(_: t.Any) -> types.ServerResult:  # noqa: ANN401
-            result = await remote_app.list_resource_templates()
+        async def _list_resource_templates(
+            req: types.ListResourceTemplatesRequest,
+        ) -> types.ServerResult:
+            result = await remote_app.list_resource_templates(params=req.params)
             return types.ServerResult(result)
 
         app.request_handlers[types.ListResourceTemplatesRequest] = _list_resource_templates
@@ -85,8 +87,10 @@
     if capabilities.tools:
         logger.debug("Capabilities: adding Tools...")
 
-        async def _list_tools(_: t.Any) -> types.ServerResult:  # noqa: ANN401
-            tools = await remote_app.list_tools()
+        async def _list_tools(req: t.Any) -> types.ServerResult:  # noqa: ANN401
+            tools = await remote_app.list_tools(
+                params=req.params if req is not None else None,
+            )
             return types.ServerResult(tools)
 
         app.request_handlers[types.ListToolsRequest] = _list_tools
